### ISSUE TITLE
comparing to old 3D Viewer

### ISSUE THREAD

**Issue by tischi** [bene51/3Dscript#3]
Is the viewer here different/better/faster than the old one that is shipped with Fiji as "3D Viewer".
Specifically, running below code was sometimes slow (several seconds) for me:
```
final long start = System.currentTimeMillis();
final Image3DUniverse univ = new Image3DUniverse();
univ.show();
final Content content = univ.addMesh( objectMask, new Color3f( 1.0f, 1.0f, 1.0f ), "object at level " + ll, 250, new boolean[]{ true, true, true }, 1 );
Utils.log( "Computed mesh and created 3D display in " + (System.currentTimeMillis() - start) + " ms" );
```
I thus wonder whether the code in this repo might be faster?

**Comment by bene51** [comment]
Hi,

the viewer here is not meant as a replacement for the old 3D Viewer. But you are right, the old viewer hit some limitations which were mainly imposed by Java3D. One such limitation was the size of the volumes that could be rendered, because the texture rendering mechanism forced a 3-fold duplication of the images.

For visualizing bigger image volumes I started out writing a purely CPU-based renderer, and didn't care about the (lack of) speed. My focus was more on making it easy to animate the data for a pleasing movie. This was a simple volume renderer for visualizing a single volume.
When I started working in the facility, more and more people wanted me to render their data, so I was actually fed up with the time it took to render, and so I implemented the actual rendering in OpenCL.

So in summary: If you want to create nice animations of single, bigger (not huge, the data still needs to fit in the GPU memory) image volumes, the viewer here is the one to use (and right now, I myself use exclusively this one). It will indeed be faster, because it handles only one volume, so some computational overhead like collision detection etc. not necessary. If you want to visualize surfaces (i.e. meshes) or multiple volumes or a combination thereof, use the old 3D Viewer.

I hope I was able to answer your question.

**Comment by tischi** [comment]
Hi,

Thank you for your answer!

I am not really into 3D visualization and thus have a technical question:

When you talk about "rendering volumes" does that basically mean maximum projections along current view-point?  I need to have a realistic 3D impression of smallish single objects. I guess for this one needs meshes and surface rendering?! 

Regarding the meshes: could I generate the mesh object _outside_ of the old 3DViewer and directly view it?

**Comment by bene51** [comment]
Have you had a look at the docs here:
[http://132.187.25.13/ij3d/?category=Documentation&page=devdoc/CustomSurface](http://132.187.25.13/ij3d/?category=Documentation&page=devdoc/CustomSurface)

here the important lines:
`
List<Point3f> mesh = new ArrayList<Point3f>();
mesh.add(new Point3f(-100, -100, 50));
mesh.add(new Point3f(100, -100, 50));
mesh.add(new Point3f(0, 100, 50));

Image3DUniverse univ = new Image3DUniverse();
univ.show();

// Add the mesh as a triangle
CustomTriangleMesh tm = new CustomTriangleMesh(mesh);
tm.setColor(new Color3f(0, 0, 1));
Content c = univ.addCustomMesh(tm, "triangle");
`

For more triangles, just add more points to the list.
